File: worker/metrics.py
from prometheus_client import Counter, Gauge, Histogram, CollectorRegistry

registry = CollectorRegistry()

songs_downloaded_counter = Counter(
    'worker_songs_downloaded_total',
    'Total number of songs downloaded',
    ['status'],
    registry=registry
)

download_duration_histogram = Histogram(
    'worker_download_duration_seconds',
    'Time taken to download a song',
    registry=registry
)

jobs_processed_counter = Counter(
    'worker_jobs_processed_total',
    'Total number of jobs processed',
    ['status'],
    registry=registry
)

rabbit_connection_status = Gauge(
    'worker_rabbit_connected',
    'RabbitMQ connection status (1=connected, 0=disconnected)',
    registry=registry
)

redis_connection_status = Gauge(
    'worker_redis_connected',
    'Redis connection status (1=connected, 0=disconnected)',
    registry=registry
)

active_downloads_gauge = Gauge(
    'worker_active_downloads',
    'Number of currently active downloads',
    registry=registry
)

job_queue_depth = Gauge(
    'worker_job_queue_depth',
    'Current depth of job queue',
    registry=registry
)

File: producer/metrics.js
const client = require('prom-client');

client.collectDefaultMetrics();

const jobsCreatedCounter = new client.Counter({
  name: 'producer_jobs_created_total',
  help: 'Total number of jobs created',
  labelNames: ['status']
});

const jobStatusCheckCounter = new client.Counter({
  name: 'producer_job_status_checks_total',
  help: 'Total number of job status checks',
  labelNames: ['status']
});

const rabbitConnectionAttempts = new client.Counter({
  name: 'producer_rabbit_connection_attempts_total',
  help: 'Total RabbitMQ connection attempts',
  labelNames: ['status']
});

const redisConnectionStatus = new client.Gauge({
  name: 'producer_redis_connected',
  help: 'Redis connection status (1=connected, 0=disconnected)'
});

const rabbitConnectionStatus = new client.Gauge({
  name: 'producer_rabbit_connected',
  help: 'RabbitMQ connection status (1=connected, 0=disconnected)'
});

const jobQueueSize = new client.Gauge({
  name: 'producer_job_queue_size',
  help: 'Current number of jobs in queue'
});

module.exports = {
  client,
  jobsCreatedCounter,
  jobStatusCheckCounter,
  rabbitConnectionAttempts,
  redisConnectionStatus,
  rabbitConnectionStatus,
  jobQueueSize,
  register: client.register
};
